## Step 1 : Read Training CSV file

In [1]:
from warnings import filterwarnings
filterwarnings('ignore')

In [2]:
import pandas as pd
df = pd.read_csv('training_set.csv',na_values=['','NA'],keep_default_na=False)
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

## Step 2 : Check Missing value

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

In [4]:
s = df.isna().sum()
s

Id                 0
MSSubClass         0
MSZoning           0
LotFrontage      259
LotArea            0
                ... 
MoSold             0
YrSold             0
SaleType           0
SaleCondition      0
SalePrice          0
Length: 81, dtype: int64

In [5]:
s[s>0]

LotFrontage      259
Alley           1369
MasVnrType         8
MasVnrArea         8
BsmtQual          37
BsmtCond          37
BsmtExposure      38
BsmtFinType1      37
BsmtFinType2      38
Electrical         1
FireplaceQu      690
GarageType        81
GarageYrBlt       81
GarageFinish      81
GarageQual        81
GarageCond        81
PoolQC          1453
Fence           1179
MiscFeature     1406
dtype: int64

In [6]:
df.duplicated().sum()

0

In [7]:
X = df.drop(columns=['Id','SalePrice'])
Y = df[['SalePrice']]

In [8]:
X.head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0          60       RL         65.0     8450   Pave   NaN      Reg   
1          20       RL         80.0     9600   Pave   NaN      Reg   
2          60       RL         68.0    11250   Pave   NaN      IR1   
3          70       RL         60.0     9550   Pave   NaN      IR1   
4          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities LotConfig  ... ScreenPorch PoolArea PoolQC Fence  \
0         Lvl    AllPub    Inside  ...           0        0    NaN   NaN   
1         Lvl    AllPub       FR2  ...           0        0    NaN   NaN   
2         Lvl    AllPub    Inside  ...           0        0    NaN   NaN   
3         Lvl    AllPub    Corner  ...           0        0    NaN   NaN   
4         Lvl    AllPub       FR2  ...           0        0    NaN   NaN   

  MiscFeature MiscVal  MoSold  YrSold  SaleType  SaleCondition  
0         NaN       0       2    2008        WD         Normal  
1         NaN       0       5    2007        WD         Normal  
2         NaN       0       9    2008        WD         Normal  
3         NaN       0       2    2006        WD        Abnorml  
4         NaN       0      12    2008        WD         Normal  

[5 rows x 79 columns]

In [9]:
Y.head()

SalePrice
0     208500
1     181500
2     223500
3     140000
4     250000

In [10]:
X.columns

Index(['MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street', 'Alley',
       'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope',
       'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle',
       'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd', 'RoofStyle',
       'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'MasVnrArea',
       'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond',
       'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1', 'BsmtFinType2',
       'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating', 'HeatingQC',
       'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF', 'LowQualFinSF',
       'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath',
       'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual', 'TotRmsAbvGrd',
       'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType', 'GarageYrBlt',
       'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual', 'GarageCond',
       'PavedDrive', 'Wo

In [11]:
X.dtypes

MSSubClass         int64
MSZoning          object
LotFrontage      float64
LotArea            int64
Street            object
                  ...   
MiscVal            int64
MoSold             int64
YrSold             int64
SaleType          object
SaleCondition     object
Length: 79, dtype: object

In [12]:
cat = list(X.columns[X.dtypes=='object'])
con = list(X.columns[X.dtypes!='object'])

In [13]:
cat

['MSZoning',
 'Street',
 'Alley',
 'LotShape',
 'LandContour',
 'Utilities',
 'LotConfig',
 'LandSlope',
 'Neighborhood',
 'Condition1',
 'Condition2',
 'BldgType',
 'HouseStyle',
 'RoofStyle',
 'RoofMatl',
 'Exterior1st',
 'Exterior2nd',
 'MasVnrType',
 'ExterQual',
 'ExterCond',
 'Foundation',
 'BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'BsmtFinType1',
 'BsmtFinType2',
 'Heating',
 'HeatingQC',
 'CentralAir',
 'Electrical',
 'KitchenQual',
 'Functional',
 'FireplaceQu',
 'GarageType',
 'GarageFinish',
 'GarageQual',
 'GarageCond',
 'PavedDrive',
 'PoolQC',
 'Fence',
 'MiscFeature',
 'SaleType',
 'SaleCondition']

In [14]:
con

['MSSubClass',
 'LotFrontage',
 'LotArea',
 'OverallQual',
 'OverallCond',
 'YearBuilt',
 'YearRemodAdd',
 'MasVnrArea',
 'BsmtFinSF1',
 'BsmtFinSF2',
 'BsmtUnfSF',
 'TotalBsmtSF',
 '1stFlrSF',
 '2ndFlrSF',
 'LowQualFinSF',
 'GrLivArea',
 'BsmtFullBath',
 'BsmtHalfBath',
 'FullBath',
 'HalfBath',
 'BedroomAbvGr',
 'KitchenAbvGr',
 'TotRmsAbvGrd',
 'Fireplaces',
 'GarageYrBlt',
 'GarageCars',
 'GarageArea',
 'WoodDeckSF',
 'OpenPorchSF',
 'EnclosedPorch',
 '3SsnPorch',
 'ScreenPorch',
 'PoolArea',
 'MiscVal',
 'MoSold',
 'YrSold']

In [15]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler,OrdinalEncoder
from sklearn.compose import ColumnTransformer


In [16]:
num_pipe1  = Pipeline(steps=[('impute',SimpleImputer(strategy='mean')),
                            ('scaler',StandardScaler())])

In [17]:
cat_pipe1 = Pipeline(steps=[('impute',SimpleImputer(strategy='constant',fill_value='Not_Avail')),
                           ('ordinal',OrdinalEncoder())])

In [18]:
pre1 = ColumnTransformer([('num',num_pipe1,con),
                         ('cat',cat_pipe1,cat)]).set_output(transform='pandas')

In [19]:
pre1

ColumnTransformer(transformers=[('num',
                                 Pipeline(steps=[('impute', SimpleImputer()),
                                                 ('scaler', StandardScaler())]),
                                 ['MSSubClass', 'LotFrontage', 'LotArea',
                                  'OverallQual', 'OverallCond', 'YearBuilt',
                                  'YearRemodAdd', 'MasVnrArea', 'BsmtFinSF1',
                                  'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF',
                                  '1stFlrSF', '2ndFlrSF', 'LowQualFinSF',
                                  'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath',
                                  '...
                                 ['MSZoning', 'Street', 'Alley', 'LotShape',
                                  'LandContour', 'Utilities', 'LotConfig',
                                  'LandSlope', 'Neighborhood', 'Condition1',
                                  'Condition2', 'BldgType', 'HouseStyle',
                                  'RoofStyle', 'RoofMatl', 'Exterior1st',
                                  'Exterior2nd', 'MasVnrType', 'ExterQual',
                                  'ExterCond', 'Foundation', 'BsmtQual',
                                  'BsmtCond', 'BsmtExposure', 'BsmtFinType1',
                                  'BsmtFinType2', 'Heating', 'HeatingQC',
                                  'CentralAir', 'Electrical', ...])])

In [20]:
X_pre = pre1.fit_transform(X)

In [21]:
X_pre.head()

num__MSSubClass  num__LotFrontage  num__LotArea  num__OverallQual  \
0         0.073375         -0.229372     -0.207142          0.651479   
1        -0.872563          0.451936     -0.091886         -0.071836   
2         0.073375         -0.093110      0.073480          0.651479   
3         0.309859         -0.456474     -0.096897          0.651479   
4         0.073375          0.633618      0.375148          1.374795   

   num__OverallCond  num__YearBuilt  num__YearRemodAdd  num__MasVnrArea  \
0         -0.517200        1.050994           0.878668         0.511418   
1          2.179628        0.156734          -0.429577        -0.574410   
2         -0.517200        0.984752           0.830215         0.323060   
3         -0.517200       -1.863632          -0.720298        -0.574410   
4         -0.517200        0.951632           0.733308         1.364570   

   num__BsmtFinSF1  num__BsmtFinSF2  ...  cat__GarageType  cat__GarageFinish  \
0         0.575425        -0.288653  ...              1.0                2.0   
1         1.171992        -0.288653  ...              1.0                2.0   
2         0.092907        -0.288653  ...              1.0                2.0   
3        -0.499274        -0.288653  ...              5.0                3.0   
4         0.463568        -0.288653  ...              1.0                2.0   

   cat__GarageQual  cat__GarageCond  cat__PavedDrive  cat__PoolQC  cat__Fence  \
0              5.0              5.0              2.0          3.0         4.0   
1              5.0              5.0              2.0          3.0         4.0   
2              5.0              5.0              2.0          3.0         4.0   
3              5.0              5.0              2.0          3.0         4.0   
4              5.0              5.0              2.0          3.0         4.0   

   cat__MiscFeature  cat__SaleType  cat__SaleCondition  
0               1.0            8.0                 4.0  
1               1.0            8.0                 4.0  
2               1.0            8.0                 4.0  
3               1.0            8.0                 0.0  
4               1.0            8.0                 4.0  

[5 rows x 79 columns]

In [22]:
from sklearn.linear_model import LinearRegression
from sklearn.feature_selection import SequentialFeatureSelector
lr = LinearRegression()
sel = SequentialFeatureSelector(estimator=lr,n_features_to_select='auto',direction='backward')
sel.fit_transform(X_pre,Y)
sel_cols = sel.get_feature_names_out()
sel_cols

array(['num__MSSubClass', 'num__LotArea', 'num__OverallQual',
       'num__OverallCond', 'num__YearBuilt', 'num__MasVnrArea',
       'num__BsmtFinSF1', 'num__BsmtFinSF2', 'num__BsmtUnfSF',
       'num__TotalBsmtSF', 'num__GrLivArea', 'num__BsmtFullBath',
       'num__KitchenAbvGr', 'num__TotRmsAbvGrd', 'num__Fireplaces',
       'num__GarageCars', 'num__WoodDeckSF', 'num__EnclosedPorch',
       'num__ScreenPorch', 'num__PoolArea', 'num__YrSold',
       'cat__LandContour', 'cat__Neighborhood', 'cat__BldgType',
       'cat__HouseStyle', 'cat__RoofMatl', 'cat__Exterior1st',
       'cat__MasVnrType', 'cat__ExterQual', 'cat__BsmtQual',
       'cat__BsmtCond', 'cat__BsmtExposure', 'cat__HeatingQC',
       'cat__KitchenQual', 'cat__Functional', 'cat__FireplaceQu',
       'cat__GarageCond', 'cat__PavedDrive', 'cat__Fence',
       'cat__MiscFeature'], dtype=object)

In [23]:
len(sel_cols)

40

In [24]:
sel_cols[0]

'num__MSSubClass'

In [25]:
sel_cols[0].split('__')

['num', 'MSSubClass']

In [26]:
imp_cols = []
for i in sel_cols:
    s = i.split('__')[1]
    imp_cols.append(s)

In [27]:
imp_cols

['MSSubClass',
 'LotArea',
 'OverallQual',
 'OverallCond',
 'YearBuilt',
 'MasVnrArea',
 'BsmtFinSF1',
 'BsmtFinSF2',
 'BsmtUnfSF',
 'TotalBsmtSF',
 'GrLivArea',
 'BsmtFullBath',
 'KitchenAbvGr',
 'TotRmsAbvGrd',
 'Fireplaces',
 'GarageCars',
 'WoodDeckSF',
 'EnclosedPorch',
 'ScreenPorch',
 'PoolArea',
 'YrSold',
 'LandContour',
 'Neighborhood',
 'BldgType',
 'HouseStyle',
 'RoofMatl',
 'Exterior1st',
 'MasVnrType',
 'ExterQual',
 'BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'HeatingQC',
 'KitchenQual',
 'Functional',
 'FireplaceQu',
 'GarageCond',
 'PavedDrive',
 'Fence',
 'MiscFeature']

In [28]:
X_sel = X[imp_cols]
X_sel.head()

MSSubClass  LotArea  OverallQual  OverallCond  YearBuilt  MasVnrArea  \
0          60     8450            7            5       2003       196.0   
1          20     9600            6            8       1976         0.0   
2          60    11250            7            5       2001       162.0   
3          70     9550            7            5       1915         0.0   
4          60    14260            8            5       2000       350.0   

   BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  ...  BsmtCond  \
0         706           0        150          856  ...        TA   
1         978           0        284         1262  ...        TA   
2         486           0        434          920  ...        TA   
3         216           0        540          756  ...        Gd   
4         655           0        490         1145  ...        TA   

   BsmtExposure  HeatingQC  KitchenQual  Functional  FireplaceQu  GarageCond  \
0            No         Ex           Gd         Typ          NaN          TA   
1            Gd         Ex           TA         Typ           TA          TA   
2            Mn         Ex           Gd         Typ           TA          TA   
3            No         Gd           Gd         Typ           Gd          TA   
4            Av         Ex           Gd         Typ           TA          TA   

   PavedDrive  Fence  MiscFeature  
0           Y    NaN          NaN  
1           Y    NaN          NaN  
2           Y    NaN          NaN  
3           Y    NaN          NaN  
4           Y    NaN          NaN  

[5 rows x 40 columns]

In [29]:
cat_sel = list(X_sel.columns[X_sel.dtypes=='object'])
con_sel = list(X_sel.columns[X_sel.dtypes!='object'])

In [30]:
cat_sel

['LandContour',
 'Neighborhood',
 'BldgType',
 'HouseStyle',
 'RoofMatl',
 'Exterior1st',
 'MasVnrType',
 'ExterQual',
 'BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'HeatingQC',
 'KitchenQual',
 'Functional',
 'FireplaceQu',
 'GarageCond',
 'PavedDrive',
 'Fence',
 'MiscFeature']

In [31]:
con_sel

['MSSubClass',
 'LotArea',
 'OverallQual',
 'OverallCond',
 'YearBuilt',
 'MasVnrArea',
 'BsmtFinSF1',
 'BsmtFinSF2',
 'BsmtUnfSF',
 'TotalBsmtSF',
 'GrLivArea',
 'BsmtFullBath',
 'KitchenAbvGr',
 'TotRmsAbvGrd',
 'Fireplaces',
 'GarageCars',
 'WoodDeckSF',
 'EnclosedPorch',
 'ScreenPorch',
 'PoolArea',
 'YrSold']

In [32]:
from sklearn.preprocessing import OneHotEncoder

In [33]:
num_pipe2 = Pipeline(steps=[('impute',SimpleImputer(strategy='mean')),
                            ('scaler',StandardScaler())])

In [34]:
cat_pipe2 = Pipeline(steps=[('impute',SimpleImputer(strategy='constant',fill_value='Not_Avail')),
                            ('ohe',OneHotEncoder(handle_unknown='ignore',sparse_output=False))])

In [35]:
pre2 = ColumnTransformer([('num',num_pipe2,con_sel),
                          ('cat',cat_pipe2,cat_sel)]).set_output(transform='pandas')

In [36]:
pre2

ColumnTransformer(transformers=[('num',
                                 Pipeline(steps=[('impute', SimpleImputer()),
                                                 ('scaler', StandardScaler())]),
                                 ['MSSubClass', 'LotArea', 'OverallQual',
                                  'OverallCond', 'YearBuilt', 'MasVnrArea',
                                  'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF',
                                  'TotalBsmtSF', 'GrLivArea', 'BsmtFullBath',
                                  'KitchenAbvGr', 'TotRmsAbvGrd', 'Fireplaces',
                                  'GarageCars', 'WoodDeckSF', 'EnclosedPorc...
                                                  SimpleImputer(fill_value='Not_Avail',
                                                                strategy='constant')),
                                                 ('ohe',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False))]),
                                 ['LandContour', 'Neighborhood', 'BldgType',
                                  'HouseStyle', 'RoofMatl', 'Exterior1st',
                                  'MasVnrType', 'ExterQual', 'BsmtQual',
                                  'BsmtCond', 'BsmtExposure', 'HeatingQC',
                                  'KitchenQual', 'Functional', 'FireplaceQu',
                                  'GarageCond', 'PavedDrive', 'Fence',
                                  'MiscFeature'])])

In [37]:
X_sel_pre = pre2.fit_transform(X_sel)
X_sel_pre.head()

num__MSSubClass  num__LotArea  num__OverallQual  num__OverallCond  \
0         0.073375     -0.207142          0.651479         -0.517200   
1        -0.872563     -0.091886         -0.071836          2.179628   
2         0.073375      0.073480          0.651479         -0.517200   
3         0.309859     -0.096897          0.651479         -0.517200   
4         0.073375      0.375148          1.374795         -0.517200   

   num__YearBuilt  num__MasVnrArea  num__BsmtFinSF1  num__BsmtFinSF2  \
0        1.050994         0.511418         0.575425        -0.288653   
1        0.156734        -0.574410         1.171992        -0.288653   
2        0.984752         0.323060         0.092907        -0.288653   
3       -1.863632        -0.574410        -0.499274        -0.288653   
4        0.951632         1.364570         0.463568        -0.288653   

   num__BsmtUnfSF  num__TotalBsmtSF  ...  cat__Fence_GdPrv  cat__Fence_GdWo  \
0       -0.944591         -0.459303  ...               0.0              0.0   
1       -0.641228          0.466465  ...               0.0              0.0   
2       -0.301643         -0.313369  ...               0.0              0.0   
3       -0.061670         -0.687324  ...               0.0              0.0   
4       -0.174865          0.199680  ...               0.0              0.0   

   cat__Fence_MnPrv  cat__Fence_MnWw  cat__Fence_Not_Avail  \
0               0.0              0.0                   1.0   
1               0.0              0.0                   1.0   
2               0.0              0.0                   1.0   
3               0.0              0.0                   1.0   
4               0.0              0.0                   1.0   

   cat__MiscFeature_Gar2  cat__MiscFeature_Not_Avail  cat__MiscFeature_Othr  \
0                    0.0                         1.0                    0.0   
1                    0.0                         1.0                    0.0   
2                    0.0                         1.0                    0.0   
3                    0.0                         1.0                    0.0   
4                    0.0                         1.0                    0.0   

   cat__MiscFeature_Shed  cat__MiscFeature_TenC  
0                    0.0                    0.0  
1                    0.0                    0.0  
2                    0.0                    0.0  
3                    0.0                    0.0  
4                    0.0                    0.0  

[5 rows x 151 columns]

In [85]:
from sklearn.model_selection import train_test_split
xtrain,xtest,ytrain,ytest = train_test_split(X_sel_pre, Y ,test_size=0.2,random_state=21)

In [86]:
xtrain.head()

num__MSSubClass  num__LotArea  num__OverallQual  num__OverallCond  \
710         -0.636078     -0.640101         -2.241782          0.381743   
1098        -0.163109     -0.452686         -1.518467          0.381743   
1286        -0.872563     -0.072844         -0.071836         -0.517200   
992          0.073375     -0.075851         -0.071836          2.179628   
631          1.492282     -0.593999          1.374795         -0.517200   

      num__YearBuilt  num__MasVnrArea  num__BsmtFinSF1  num__BsmtFinSF2  \
710        -1.201217        -0.574410        -0.973018        -0.288653   
1098       -1.168096        -0.574410         0.500854        -0.288653   
1286       -0.273836         1.924104         0.274948         0.213629   
992        -0.240715         1.309170         0.202570         0.436865   
631         1.150356         0.023903        -0.920380        -0.288653   

      num__BsmtUnfSF  num__TotalBsmtSF  ...  cat__Fence_GdPrv  \
710        -0.672923         -1.795509  ...               0.0   
1098       -1.284176         -0.878862  ...               0.0   
1286        0.250749          0.616959  ...               0.0   
992        -0.901577         -0.536830  ...               0.0   
631         2.179592          1.132288  ...               0.0   

      cat__Fence_GdWo  cat__Fence_MnPrv  cat__Fence_MnWw  \
710               0.0               0.0              0.0   
1098              0.0               0.0              0.0   
1286              0.0               0.0              0.0   
992               0.0               0.0              1.0   
631               0.0               0.0              0.0   

      cat__Fence_Not_Avail  cat__MiscFeature_Gar2  cat__MiscFeature_Not_Avail  \
710                    1.0                    0.0                         1.0   
1098                   1.0                    0.0                         1.0   
1286                   1.0                    0.0                         1.0   
992                    0.0                    0.0                         1.0   
631                    1.0                    0.0                         1.0   

      cat__MiscFeature_Othr  cat__MiscFeature_Shed  cat__MiscFeature_TenC  
710                     0.0                    0.0                    0.0  
1098                    0.0                    0.0                    0.0  
1286                    0.0                    0.0                    0.0  
992                     0.0                    0.0                    0.0  
631                     0.0                    0.0                    0.0  

[5 rows x 151 columns]

In [87]:
ytrain.head()

SalePrice
710       52000
1098     128000
1286     143000
992      187000
631      209500

In [88]:
xtest.head()

num__MSSubClass  num__LotArea  num__OverallQual  num__OverallCond  \
880         -0.872563     -0.350058         -0.795151         -0.517200   
605          0.073375      0.309002          0.651479          0.381743   
1166        -0.872563     -0.004192          1.374795         -0.517200   
216         -0.872563     -0.207142          0.651479         -0.517200   
970         -0.163109      0.028380         -1.518467         -1.416142   

      num__YearBuilt  num__MasVnrArea  num__BsmtFinSF1  num__BsmtFinSF2  \
880         1.117235        -0.574410         1.176379        -0.288653   
605        -0.207594         0.400620         0.022723        -0.288653   
1166        1.216598        -0.175535        -0.973018        -0.288653   
216         1.084115         0.899214         1.101808        -0.288653   
970        -0.737526        -0.574410        -0.973018        -0.288653   

      num__BsmtUnfSF  num__TotalBsmtSF  ...  cat__Fence_GdPrv  \
880        -1.035147          0.074268  ...               0.0   
605        -0.573311         -0.659961  ...               0.0   
1166        2.550871          1.451518  ...               0.0   
216        -0.174865          0.863222  ...               0.0   
970         0.345832         -0.769412  ...               0.0   

      cat__Fence_GdWo  cat__Fence_MnPrv  cat__Fence_MnWw  \
880               0.0               0.0              0.0   
605               0.0               0.0              0.0   
1166              0.0               0.0              0.0   
216               0.0               0.0              0.0   
970               0.0               0.0              0.0   

      cat__Fence_Not_Avail  cat__MiscFeature_Gar2  cat__MiscFeature_Not_Avail  \
880                    1.0                    0.0                         1.0   
605                    1.0                    0.0                         1.0   
1166                   1.0                    0.0                         1.0   
216                    1.0                    0.0                         1.0   
970                    1.0                    0.0                         1.0   

      cat__MiscFeature_Othr  cat__MiscFeature_Shed  cat__MiscFeature_TenC  
880                     0.0                    0.0                    0.0  
605                     0.0                    0.0                    0.0  
1166                    0.0                    0.0                    0.0  
216                     0.0                    0.0                    0.0  
970                     0.0                    0.0                    0.0  

[5 rows x 151 columns]

In [89]:
ytest.head()

SalePrice
880      157000
605      205000
1166     245350
216      210000
970      135000

In [90]:
model1 = LinearRegression()
model1.fit(xtrain,ytrain)

LinearRegression()

In [91]:
model1.score(xtrain,ytrain)

0.9253879181006113

In [92]:
model1.score(xtest,ytest)

0.8173162445726133

In [93]:
from sklearn.linear_model import Ridge 
model2 = Ridge(alpha=1)
model2.fit(xtrain, ytrain)

Ridge(alpha=1)

In [94]:
model2.score(xtrain,ytrain)

0.911076053186626

In [95]:
model2.score(xtest,ytest)

0.838894820260339

In [96]:
def evalute_model(x,y):
    ypred = model2.predict(x)
    from sklearn.metrics import mean_squared_error,mean_absolute_error,mean_absolute_percentage_error,r2_score
    mse = mean_squared_error(y,ypred)
    rmse = mse**(1/2)
    mae = mean_absolute_error(y,ypred)
    mape = mean_absolute_percentage_error(y,ypred)
    r2 = r2_score(y,ypred)
    print(f'MSE:{mse:.2f}')
    print(f'RMSE:{rmse:.2f}')
    print(f'MAE :{mae:.2f}')
    print(f'MAPE : {mape:.4f}')
    print(f'R2:{r2:.4f}')

In [97]:
evalute_model(xtrain,ytrain)

MSE:550343683.90
RMSE:23459.41
MAE :15332.02
MAPE : 0.0912
R2:0.9111


In [98]:
evalute_model(xtest,ytest)

MSE:1091860065.19
RMSE:33043.31
MAE :17290.11
MAPE : 0.1024
R2:0.8389


In [99]:
from sklearn.linear_model import Lasso
model3 = Lasso(alpha=1)
model3.fit(xtrain,ytrain)

Lasso(alpha=1)

In [100]:
model3.score(xtrain,ytrain)

0.9253826314241478

In [101]:
model3.score(xtest,ytest)

0.8179970236744609

In [102]:
def evalute_model(x,y):
    ypred = model3.predict(x)
    from sklearn.metrics import mean_squared_error,mean_absolute_error,mean_absolute_percentage_error,r2_score
    mse = mean_squared_error(y,ypred)
    rmse = mse**(1/2)
    mae = mean_absolute_error(y,ypred)
    mape = mean_absolute_percentage_error(y,ypred)
    r2 = r2_score(y,ypred)
    print(f'MSE:{mse:.2f}')
    print(f'RMSE:{rmse:.2f}')
    print(f'MAE :{mae:.2f}')
    print(f'MAPE : {mape:.4f}')
    print(f'R2:{r2:.4f}')

In [103]:
evalute_model(xtrain,ytrain)

MSE:461801336.72
RMSE:21489.56
MAE :14672.34
MAPE : 0.0882
R2:0.9254


In [104]:
evalute_model(xtest,ytest)

MSE:1233490952.41
RMSE:35121.09
MAE :17329.20
MAPE : 0.1029
R2:0.8180


In [114]:
import numpy as np 
params = {'alpha': np.arange(start=1, stop=1000, step=1)}

In [115]:
params

{'alpha': array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
         14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
         27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
         40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
         53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
         66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
         79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
         92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103, 104,
        105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117,
        118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130,
        131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143,
        144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156,
        157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169,
        170, 171, 172, 173, 174, 175, 176,

In [116]:
## Tuning for ridge
from sklearn.model_selection import GridSearchCV
rr = Ridge()
gscv1 = GridSearchCV(estimator=rr, param_grid=params, cv=5, scoring='neg_mean_squared_error')
gscv1.fit(xtrain, ytrain)

GridSearchCV(cv=5, estimator=Ridge(),
             param_grid={'alpha': array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99...
       911, 912, 913, 914, 915, 916, 917, 918, 919, 920, 921, 922, 923,
       924, 925, 926, 927, 928, 929, 930, 931, 932, 933, 934, 935, 936,
       937, 938, 939, 940, 941, 942, 943, 944, 945, 946, 947, 948, 949,
       950, 951, 952, 953, 954, 955, 956, 957, 958, 959, 960, 961, 962,
       963, 964, 965, 966, 967, 968, 969, 970, 971, 972, 973, 974, 975,
       976, 977, 978, 979, 980, 981, 982, 983, 984, 985, 986, 987, 988,
       989, 990, 991, 992, 993, 994, 995, 996, 997, 998, 999])},
             scoring='neg_mean_squared_error')

In [117]:
gscv1.best_params_

{'alpha': 23}

In [118]:
gscv1.best_score_

-1050947982.1320308

In [119]:
best_ridge = gscv1.best_estimator_

In [120]:
best_ridge

Ridge(alpha=23)

In [121]:
best_ridge.score(xtrain,ytrain)

0.8832966136921568

In [122]:
best_ridge.score(xtest,ytest)

0.8318747056866947

In [123]:
## Tuning for Lasso
ls = Lasso()
gscv2 = GridSearchCV(estimator=ls, param_grid=params, cv=5, scoring='neg_mean_squared_error')
gscv2.fit(xtrain, ytrain)

GridSearchCV(cv=5, estimator=Lasso(),
             param_grid={'alpha': array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99...
       911, 912, 913, 914, 915, 916, 917, 918, 919, 920, 921, 922, 923,
       924, 925, 926, 927, 928, 929, 930, 931, 932, 933, 934, 935, 936,
       937, 938, 939, 940, 941, 942, 943, 944, 945, 946, 947, 948, 949,
       950, 951, 952, 953, 954, 955, 956, 957, 958, 959, 960, 961, 962,
       963, 964, 965, 966, 967, 968, 969, 970, 971, 972, 973, 974, 975,
       976, 977, 978, 979, 980, 981, 982, 983, 984, 985, 986, 987, 988,
       989, 990, 991, 992, 993, 994, 995, 996, 997, 998, 999])},
             scoring='neg_mean_squared_error')

In [124]:
gscv2.best_params_

{'alpha': 124}

In [125]:
gscv2.best_score_

-968947732.3824869

In [126]:
best_lasso = gscv2.best_estimator_

In [127]:
best_lasso

Lasso(alpha=124)

In [128]:
best_lasso.score(xtrain,ytrain)

0.913980578643248

In [129]:
best_lasso.score(xtest,ytest)

0.8262012764568645

In [130]:
from sklearn.model_selection import cross_val_score
scores_lasso = cross_val_score(best_lasso, xtrain, ytrain, cv=5, scoring='r2')
scores_lasso

array([0.62621749, 0.87094186, 0.91114986, 0.89042078, 0.9318967 ])

In [131]:
scores_lasso.mean()

0.8461253361966101

In [132]:
scores_ridge = cross_val_score(best_ridge, xtrain, ytrain, cv=5, scoring='r2')
scores_ridge

array([0.61158537, 0.85035103, 0.90102389, 0.87782459, 0.9236633 ])

In [133]:
scores_ridge.mean()

0.832889637588139

### From above model choose Best Lasso model as it has higest test r2 score of 0.8461

In [134]:
xnew = pd.read_csv('sample_set.csv',na_values=['','NA'],keep_default_na=False)
xnew.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0  1461          20       RH         80.0    11622   Pave   NaN      Reg   
1  1462          20       RL         81.0    14267   Pave   NaN      IR1   
2  1463          60       RL         74.0    13830   Pave   NaN      IR1   
3  1464          60       RL         78.0     9978   Pave   NaN      IR1   
4  1465         120       RL         43.0     5005   Pave   NaN      IR1   

  LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence MiscFeature  \
0         Lvl    AllPub  ...         120        0    NaN  MnPrv         NaN   
1         Lvl    AllPub  ...           0        0    NaN    NaN        Gar2   
2         Lvl    AllPub  ...           0        0    NaN  MnPrv         NaN   
3         Lvl    AllPub  ...           0        0    NaN    NaN         NaN   
4         HLS    AllPub  ...         144        0    NaN    NaN         NaN   

  MiscVal MoSold  YrSold  SaleType  SaleCondition  
0       0      6    2010        WD         Normal  
1   12500      6    2010        WD         Normal  
2       0      3    2010        WD         Normal  
3       0      6    2010        WD         Normal  
4       0      1    2010        WD         Normal  

[5 rows x 80 columns]

In [135]:
xnew_pre = pre2.transform(xnew)
xnew_pre.head()

num__MSSubClass  num__LotArea  num__OverallQual  num__OverallCond  \
0        -0.872563      0.110763         -0.795151          0.381743   
1        -0.872563      0.375850         -0.071836          0.381743   
2         0.073375      0.332053         -0.795151         -0.517200   
3         0.073375     -0.054002         -0.071836          0.381743   
4         1.492282     -0.552407          1.374795         -0.517200   

   num__YearBuilt  num__MasVnrArea  num__BsmtFinSF1  num__BsmtFinSF2  \
0       -0.340077        -0.574410         0.053428         0.604293   
1       -0.439440         0.023903         1.051363        -0.288653   
2        0.852269        -0.574410         0.761852        -0.288653   
3        0.885390        -0.463612         0.347326        -0.288653   
4        0.686666        -0.574410        -0.396190        -0.288653   

   num__BsmtUnfSF  num__TotalBsmtSF  ...  cat__Fence_GdPrv  cat__Fence_GdWo  \
0       -0.672923         -0.400017  ...               0.0              0.0   
1       -0.365032          0.619239  ...               0.0              0.0   
2       -0.974021         -0.295127  ...               0.0              0.0   
3       -0.550672         -0.299687  ...               0.0              0.0   
4        1.018211          0.507509  ...               0.0              0.0   

   cat__Fence_MnPrv  cat__Fence_MnWw  cat__Fence_Not_Avail  \
0               1.0              0.0                   0.0   
1               0.0              0.0                   1.0   
2               1.0              0.0                   0.0   
3               0.0              0.0                   1.0   
4               0.0              0.0                   1.0   

   cat__MiscFeature_Gar2  cat__MiscFeature_Not_Avail  cat__MiscFeature_Othr  \
0                    0.0                         1.0                    0.0   
1                    1.0                         0.0                    0.0   
2                    0.0                         1.0                    0.0   
3                    0.0                         1.0                    0.0   
4                    0.0                         1.0                    0.0   

   cat__MiscFeature_Shed  cat__MiscFeature_TenC  
0                    0.0                    0.0  
1                    0.0                    0.0  
2                    0.0                    0.0  
3                    0.0                    0.0  
4                    0.0                    0.0  

[5 rows x 151 columns]

In [136]:
preds = best_lasso.predict(xnew_pre)
preds

array([117937.96252445, 155094.21683605, 180489.75492695, ...,
       175259.65775879, 113070.55999614, 221412.14452794])

In [137]:
xnew['SalePricePred'] = preds

In [138]:
xnew.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0  1461          20       RH         80.0    11622   Pave   NaN      Reg   
1  1462          20       RL         81.0    14267   Pave   NaN      IR1   
2  1463          60       RL         74.0    13830   Pave   NaN      IR1   
3  1464          60       RL         78.0     9978   Pave   NaN      IR1   
4  1465         120       RL         43.0     5005   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0         Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
1         Lvl    AllPub  ...        0    NaN    NaN        Gar2   12500   
2         Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
3         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
4         HLS    AllPub  ...        0    NaN    NaN         NaN       0   

  MoSold YrSold  SaleType  SaleCondition  SalePricePred  
0      6   2010        WD         Normal  117937.962524  
1      6   2010        WD         Normal  155094.216836  
2      3   2010        WD         Normal  180489.754927  
3      6   2010        WD         Normal  187880.296838  
4      1   2010        WD         Normal  199653.696366  

[5 rows x 81 columns]

In [139]:
b = xnew[['Id','SalePricePred']]
b.head()

Id  SalePricePred
0  1461  117937.962524
1  1462  155094.216836
2  1463  180489.754927
3  1464  187880.296838
4  1465  199653.696366

In [140]:
b.to_csv('Backward_result',index=False)